src/Symfony/Component/DependencyInjection/Loader/XmlFileLoader.php

```diff
@@ -315,8 +315,12 @@ private function parseDefinition(\DOMElement $service, string $file, Definition
         $tags = $this->getChildren($service, 'tag');
 
         foreach ($tags as $tag) {
-            $parameters = [];
             $tagName = $tag->nodeValue;
+            if ('' === $tagName && '' === $tagName = $tag->getAttribute('name')) {
+                throw new InvalidArgumentException(sprintf('The tag name for service "%s" in "%s" must be a non-empty string.', (string) $service->getAttribute('id'), $file));
+            }
+
+            $parameters = $this->getRecursiveTagAttributes($tag, sprintf('The attribute name of tag "%s" for service "%s" in %s must be a non-empty string.', $tagName, (string) $service->getAttribute('id'), $file));
             foreach ($tag->attributes as $name => $node) {
```

Comment:
I think something is missing here, that would allow removing this foreach. Right now, this looks like duplication with the new method.

Reply:
The foreach loop is to handle attributes on the tag itself, the method handles nested arguments.

Example in the initial comment:
```xml
<?xml version="1.0" ?>
<container xmlns="http://symfony.com/schema/dic/services"
    xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
    xsi:schemaLocation="http://symfony.com/schema/dic/services https://symfony.com/schema/dic/services/services-1.0.xsd">
  <services>
    <service id="foo" class="BarClass" public="true">
        <tag name="foo_tag" some-option="cat">
            <attribute name="array_option">
                <attribute name="key1">lorem</attribute>
                <attribute name="key2">ipsum</attribute>
            </attribute>
        </tag>
    </service>
  </services>
</container>
```

The foreach would read `some-option` while the method recursively reads `<atttribute …>`